src/Sentry/Internal/SystemDiagnosticsMetricsListener.cs

```diff
@@ -0,0 +1,86 @@
+#if NET8_0_OR_GREATER
+using System.Diagnostics.Metrics;
+
+namespace Sentry.Internal;
+
+internal class SystemDiagnosticsMetricsListener : IDisposable
+{
+    private readonly Lazy<IMetricAggregator> _metricsAggregator;
+    private IMetricAggregator MetricsAggregator => _metricsAggregator.Value;
+    internal static SystemDiagnosticsMetricsListener? DefaultListener;
+
+    internal readonly MeterListener _sentryListener = new ();
+
+    public SystemDiagnosticsMetricsListener(IEnumerable<SubstringOrRegexPattern> captureInstruments)
+        : this(captureInstruments, () => SentrySdk.Metrics)
+    {
+    }
+
+    /// <summary>
+    /// Overload for testing purposes - allows us to supply a mock IMetricAggregator
+    /// </summary>
+    internal SystemDiagnosticsMetricsListener(IEnumerable<SubstringOrRegexPattern> captureInstruments, Func<IMetricAggregator> metricsAggregatorResolver)
+    {
+        _metricsAggregator = new Lazy<IMetricAggregator>(metricsAggregatorResolver);
+        _sentryListener.InstrumentPublished = (instrument, listener) =>
+        {
+            if (captureInstruments!.Any(x => x.IsMatch(instrument.Name)))
+            {
+                listener.EnableMeasurementEvents(instrument);
+            }
+        };
+        _sentryListener.SetMeasurementEventCallback<byte>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<int>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<float>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<decimal>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<short>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<long>(RecordMeasurement);
+        _sentryListener.SetMeasurementEventCallback<double>(RecordMeasurement);
+        _sentryListener.Start();
+    }
+
+    internal static void InitializeDefaultListener(IEnumerable<SubstringOrRegexPattern> captureInstruments)
+    {
+        var oldListener = Interlocked.Exchange(
+            ref DefaultListener,
+            new SystemDiagnosticsMetricsListener(captureInstruments)
+            );
+        oldListener?.Dispose();
+    }
+
+    internal void RecordMeasurement<T>(
+        Instrument instrument,
+        T measurement,
+        ReadOnlySpan<KeyValuePair<string, object?>> tags,
+        object? _)
+        where T: struct
+    {
+        var unit = MeasurementUnit.Parse(instrument.Unit);
+        var tagDict = tags.ToImmutableArray().ToImmutableDictionary(
```

Comment:
👁️ it's almost midnight I need to get back to this and do a proper review 😅

Reply:
Worth profiling this to see the impact. This being ReadOnlySpan<T> and us allocating immutableArray then dictionary from it seems like could show up depending how often this is called.

Reply:
I've added a separate issue to track this:
- https://github.com/getsentry/sentry-dotnet/issues/3093